nbgitpuller/pull.py

```diff
@@ -88,58 +71,98 @@ def _initialize_repo(self):
-        yield from self._pull_and_resolve_conflicts()
 
-    def _reset_deleted_files(self):
+    def reset_deleted_files(self):
         """
         Runs the equivalent of git checkout -- <file> for each file that was
         deleted. This allows us to delete a file, hit an interact link, then get a
         clean version of the file again.
         """
 
-        status = subprocess.check_output(['git', 'status'], cwd=self.repo_dir)
-        deleted_files = self.DELETED_FILE_REGEX.findall(status.decode('utf-8'))
+        deleted_files = subprocess.check_output([
+            'git', 'ls-files', '--deleted'
+        ], cwd=self.repo_dir).decode().strip().split('\n')
 
         for filename in deleted_files:
-            yield from execute_cmd(['git', 'checkout', '--', filename], cwd=self.repo_dir)
-            logging.info('Resetted {}'.format(filename))
+            if filename:  # Filter out empty lines
+                yield from execute_cmd(['git', 'checkout', '--', filename], cwd=self.repo_dir)
 
-    def _make_commit(self):
+    def repo_is_dirty(self):
         """
-        Commit local changes
+        Return true if repo is dirty
         """
-        if self.repo_is_dirty():
-            yield from execute_cmd(['git', 'checkout', self.branch_name], cwd=self.repo_dir)
-            yield from execute_cmd(['git', 'commit', '-am', 'WIP'], cwd=self.repo_dir)
-            logging.info('Made WIP commit')
+        try:
+            subprocess.check_call(['git', 'diff-files', '--quiet'], cwd=self.repo_dir)
+            # Return code is 0
+            return False
+        except subprocess.CalledProcessError:
+            return True
+
+    def update_remotes(self):
+        """
+        Do a git fetch so our remotes are up to date
+        """
+        yield from execute_cmd(['git', 'fetch'])
 
-    def _pull_and_resolve_conflicts(self):
+    def find_upstream_changed(self, kind):
         """
-        Git pulls, resolving conflicts with -Xours
+        Return list of files that have been changed upstream belonging to a particular kind of change
         """
+        output = subprocess.check_output([
+            'git', 'log', '{}..origin/{}'.format(self.branch_name, self.branch_name),
+            '--oneline', '--name-status'
+        ], cwd=self.repo_dir).decode()
+        files = []
+        for line in output.split('\n'):
+            if line.startswith(kind):
+                files.append(os.path.join(self.repo_dir, line.split('\t', 1)[1]))
+
+        return files
+
+    def rename_local_untracked(self):
+        """
+        Rename local untracked files that would require pulls
+        """
+        # Find what files have been added!
+        new_upstream_files = self.find_upstream_changed('A')
+        for f in new_upstream_files:
+            if os.path.exists(f):
+                # If there's a file extension, put the timestamp before that
+                ts = datetime.datetime.now().strftime('__%Y%m%d%H%M%S')
```

Comment:
so I think here Fernando's extra idea was to use this logic when pulling modified binary files as well